src/librustc/ty/structural_match.rs

```diff
@@ -207,3 +301,186 @@ impl<'a, 'tcx> TypeVisitor<'tcx> for Search<'a, 'tcx> {
+
+        // Note: Cannot use register_bound here, because it requires (but does
+        // not check) that the given trait has no type parameters apart from
+        // `Self`, but `PartialEq` has a type parameter that defaults to `Self`.
+        let trait_ref = ty::TraitRef {
+            def_id: partial_eq_def_id,
+            substs: self.infcx.tcx.mk_substs_trait(ty, &[ty.into()]),
+        };
+        let obligation = traits::Obligation {
+            cause,
+            recursion_depth: 0,
+            param_env: self.param_env,
+            predicate: trait_ref.to_predicate(),
+        };
+        self.partialeq_fulfillment_cx.register_predicate_obligation(&self.infcx, obligation);
+    }
+}
+
+impl<'a, 'v, 'tcx> Visitor<'v> for SearchHirExpr<'a, 'tcx> {
+    fn nested_visit_map<'this>(&'this mut self) -> NestedVisitorMap<'this, 'v>
+    {
+        NestedVisitorMap::None
+    }
+    fn visit_expr(&mut self, ex: &'v hir::Expr) {
+        // When we inspect the expression on a RHS of a const, we may be able to
+        // prove that it can participate in a match pattern, without worrying
+        // about its actual type being `PartialEq`.
+        //
+        // In an ideal world we would not bother with side-stepping the
+        // `PartialEq` and just unconditionally check that the types of a
+        // constant in a match pattern are all `PartialEq`. However, one big
+        // exception is types with `for <'a> fn(...)` in them; today these are
+        // not `PartialEq` due to technical limitations.
+        //
+        // So we cut the gordian knot here: the expression analysis can allow
+        // the omission of the `PartialEq` check.
+        let mut impose_partialeq_bound = true;
+
+        let ty = self.body_tables.expr_ty(ex);
+        match &ex.kind {
+            hir::ExprKind::Struct(..) => {
+                // register structural requirement ...
+                self.register_structural_bound(ty);
+                // ... and continue expression-based traversal
+                intravisit::walk_expr(self, ex)
+            }
+
+            hir::ExprKind::Path(qpath) => {
+                let res = self.body_tables.qpath_res(qpath, ex.hir_id);
+                match res {
+                    Res::Def(DefKind::Const, def_id) |
+                    Res::Def(DefKind::AssocConst, def_id) => {
+                        let substs = self.body_tables.node_substs(ex.hir_id);
+                        if let Some(instance) = ty::Instance::resolve(
+                            self.infcx.tcx, self.param_env, def_id, substs)
+                        {
+                            let const_def_id = instance.def_id();
+                            if const_def_id.is_local() {
+                                self.recur_into_const_rhs(const_def_id);
+                            } else {
+                                // abstraction barrer for non-local definition;
+                                // traverse `typeof(expr)` instead.
```

Comment:
`mir_const_qualif` gets around this by encoding the relevant bits (literally, 2 bits) cross-crate, perhaps something similar could be done here?

You could almost just use `mir_const_qualif` directly (and add the analysis to it), but with everything being in flux that's terrible advice for now.

Reply:
I'll file a bug suggesting this change for longer-term.